### PR TITLE
Add Secrets Manager interface

### PR DESCRIPTION
**Pull request by wjordan**
This PR implements a new process for storing and using secrets within our application. It builds on top of #28877 environment-specific config by adding support for a `!Secret` tag. This tag marks certain keys as Secrets to be lazy-loaded from the [AWS Secrets Manager service](https://aws.amazon.com/secrets-manager/), through calls to the [`GetSecretValue` API](https://docs.aws.amazon.com/secretsmanager/latest/apireference/API_GetSecretValue.html).

### Referencing a secret
`!Secret`-tagged config keys (e.g.: `my_secret: !Secret`) are all mapped onto the `CDO` application-config namespace as usual, so you can reference a secret using `CDO.my_secret` in application code.

* Secrets are lazy-loaded, so the call to `GetSecretValue` will only take place when the secret is referenced.
  * An exception is running the application-server (in non-`development` environments), where all secrets are explicitly eager-loaded (before the application-server forks multiple processes), and will fail with an exception if the API fails or any expected secrets are not found.
* Secrets are cached in-memory once fetched to ensure only one API call per secret when the application is running.
* Secrets are stored and fetched separately based on the environment, so referencing `CDO.my_secret` actually fetches the secret named `staging/cdo/my_secret`, `production/cdo/my_secret`, etc depending on the environment.
* IAM permissions are tuned so that the application has read-only access to secrets in its own environment (e.g., a `staging` application can't read `production` secrets).
  * The `Developer` role is also prevented from reading secrets outside of the `development` environment for a better security posture.

### Creating/updating a secret
* Create a `config/secrets.yml` file (see `config/secrets.yml.template` as a reference) containing the configuration for the secrets you wish to create/update, then run the [`bin/update_secrets`](https://github.com/code-dot-org/code-dot-org/compare/aws-secrets?expand=1#diff-4d2a4a437c6601ff6afca597782f39b4) helper script to apply the changes.
* Update the application configuration files (`config.yml.erb` / `config/[env].yml.erb`) with `!Secret` tags for each environment where the secret now exists.
* Remember that all secrets (except those in `development`) are write-only by developers for security reasons- you shouldn't ever need to read a secret's actual value once written (leave that to the environment-specific application).

### Existing (Chef) secrets migration
A big part of the effort involved in this PR was manually migrating the existing secrets that existed across several Chef environment/role attributes, separating out the unused/legacy and non-secret configuration values, and mapping 'shared' secrets to environment-specific config. This was the source of the majority of added non-secret config-constants in this PR.
* After this PR is merged the existing Chef secrets (via `globals.yml`) will still take priority over the new Secrets Manager secrets, so the behavior should be unchanged. This will allow us to manually/carefully test and verify the secrets are working correctly across all environments by incrementally removing Chef-managed secrets while monitoring the application closely across CI builds/restarts.

---
Earlier draft discussion of this feature is in #27554.

### CODE REVIEW

File: aws/cloudformation/bootstrap_chef_stack.sh.erb
Comment (islemaster) [inline aws/cloudformation/bootstrap_chef_stack.sh.erb]: FYI @breville this change interacts with https://github.com/code-dot-org/code-dot-org/pull/29936.

Comment (wjordan) [reply aws/cloudformation/bootstrap_chef_stack.sh.erb]: good call-out, I'll be sure to coordinate on this. Getting this config-property removed entirely would simplify things and reduce the risk involved in this change, so I'm inclined to wait until the removal gets merged first.

**Review by uponthesun — COMMENTED** [review]
Looks really good! two suggestions:

- Could you add a README file in an appropriate place (maybe config/) with the documentation from this PR?
- Could you try adding tests to cover the uncovered lines from the report, within reason? https://codecov.io/gh/code-dot-org/code-dot-org/pull/30036/diff?src=pr&el=tree-more

**Comment by wjordan** [comment]
> Could you add a README file in an appropriate place (maybe config/) with the documentation from this PR?

Done (f7242f1acdcc74d3ccd3251cb9e383eed274e2ac)

> Could you try adding tests to cover the uncovered lines from the report, within reason?

Done (62850c0695af5de1a4ec7eae6d65915a7cb17de0) - though how exactly to poke Codecov to actually process the updated coverage correctly is still a mystery to me.

**Review by uponthesun — APPROVED** [review]
Yea, not sure why codecov didn't update the comment, will keep looking into that. Thanks!

**Pull request merged**
